Add stateless public plugin

kingherc opened the pull request:
Migrates the basic stateless setting, some basic checks, and some basic integration test functionality to the public elasticsearch repository. Behind an enterprise license check.

Relates ES-13599

kingherc commented on x-pack/plugin/xstateless/build.gradle:
```diff
@@ -0,0 +1,28 @@
+apply plugin: 'elasticsearch.internal-cluster-test'
+apply plugin: 'elasticsearch.internal-test-artifact'
+esplugin {
+  name = 'xstateless'
```
note. We need a different name than `stateless` to differentiate this new plugin in the public repository. Especially a different name for the folder under `x-pack`. Else the same name `stateless` would create difficulties to compile and have things working in serverless. Seems like current gradle/kotlin expects plugins to have unique names, if I'm not mistaken.

lukewhiting reviewed (approved):
One question about the testing of exceptions but otherwise looks like a great start to the project!

kingherc committed:
Add stateless public x-pack plugin

Migrates the basic stateless setting, some basic checks, and some
basic integration test functionality to the public elasticsearch
repository. Behind an enterprise license check.

Relates ES-13599

kingherc committed:
Merge branch 'main' into non-issue/ES-13599-migrate-stateless-public

kingherc committed:
Merge commit '21d031b' into non-issue/ES-13599-migrate-stateless-public

kingherc committed:
Merge remote-tracking branch 'kingherc/main' into non-issue/ES-13599-migrate-stateless-public

mosche reviewed (approved):
lgtm

kingherc committed:
Merge remote-tracking branch 'kingherc/main' into non-issue/ES-13599-migrate-stateless-public
